Update to net7.0

BenVillalobos opened the pull request:
Fixes #

### Context


### Changes Made


### Testing


### Notes
This will be chipped away at over time.

rainersigwald commented:
I would think about updating Arcade first, then our TF.

lbussell commented:
Hey @rainersigwald, @BenVillalobos, this will be needed for source-build in 7.0. I was testing this locally before I saw this PR, I am hitting the same issues as the CI here. Is there anything I can do to help move this along?

BenVillalobos commented:
@lbussell We're in early stages of working on this (Read: I'm poking at this when I have time 🙃). If you'd like to poke at it and offer some commits we can cherry-pick onto this branch, you're more than welcome! Feel free to PM me directly on this

BenVillalobos commented:
@lbussell I cherry-picked your commit to resolve the future issues you were seeing. The "This type is not part of the declared API" wasn't showing up for me. Typically when those come up (which they shouldn't here?) you just do the VS suggested action of adding them to some Public.Shipped.API.txt file. Good news is I got a green build after that! Pushing it up to see if tests fail.

As for the issue that was nowarn'd around, it looks like the deprecations to AssemblyName are going to require us to rethink our use of `AssemblyName`. It looks like we use `AssemblyName` to gather information about the generated binaries. https://github.com/dotnet/runtime/issues/59061#issuecomment-922392861 suggests looking into the `System.Reflection.Metadata` namespace for gathering assembly info. Even if we could do that, it looks like we shouldn't use AssemblyName anymore, or we may need to create our own type. 

The main culprits that need this type look like `ResolveAssemblyReference` and `BinaryTranslator`.

Based on my (very) brief investigation, I don't think working around `AssemblyName` is a reasonable thing to get volunteer work on, though we appreciate the offer! 😭 

Realistically, I suspect we'll be able to dig in more early next week.

BenVillalobos commented:
Not sure what's going on with CI, I don't see these errors locally

BenVillalobos commented:
Thinking about this some. Our build defines the netstandard/netcore public shipped API txt files to be the same. Now that we're moving to net7.0, technically none of the API's in "netstandard" have been shipped yet because we haven't shipped these API's in net7.0.

Not 100% sure this is what's happening under the hood, but I'm going to clear out public api's shipped txt under netstandard and have VS fix it back into unshipped.

rainersigwald commented:
That _shouldn't_ be the problem but heck if I know what the actual problem is (doesn't repro on my machine either).

BenVillalobos commented:
Specifically removing the API's causing errors and having VS resolve them doesn't fix it.

Marking functions as public "because the analyzer didn't see it" doesn't resolve it either

BenVillalobos commented:
@sharwell any guidance here? API's are clearly public, but the analyzers are complaining about them. Maybe I'm missing something

Forgind commented:
Preventing the warning from AssemblyName's ProcessorArchitecture being deprecated will make our caches not deserialize properly. Solution: delete the cache once (or just build again, since it should just be a warning), and it'll work the second time.

BenVillalobos commented:
@Forgind 's commits for future reference: https://github.com/forgind/msbuild/tree/net7.0-here-we-go

BenVillalobos commented:
> @sharwell any guidance here? API's are clearly public, but the analyzers are complaining about them. Maybe I'm missing something

@jaredpar I just realized @sharlwell is oof, any ideas?

jaredpar commented:
You're running afoul of a bug in the new `ref` safety rules. It is upset about this constructor:

https://github.com/dotnet/msbuild/blob/3a92cefaf053c9377037d0af0c67736867cc9711/src/StringTools/InternableString.cs#L36

It believes the ctor may capture the parameter by `ref`. This is actually a bug in the current toolset. The RC SDK will compile this without error. Short term you can mark the reference as `scoped ref` and it will fix the problem.

AraHaan commented:
Why use AssemblyName over Assembly.GetName().Name or .FullName?

BenVillalobos commented:
@AraHaan /shrug. We could change that, but there's still the task of obtaining the processor architecture, which we're exploring how to do in a reasonable way: https://github.com/dotnet/runtime/issues/74040

AraHaan commented:
> @AraHaan /shrug. We could change that, but there's still the task of obtaining the processor architecture, which we're exploring how to do in a reasonable way: [dotnet/runtime#74040](https://github.com/dotnet/runtime/issues/74040)

So basically getting the architecture of the assembly itself + the .NET Host (to compare and prevent loading the wrong architecture in the host architecture)?

BenVillalobos commented:
I've simplified things a bit for future generations: Most places that needed a manual change from `6` to `7` was replaced with a single property. I also added a comment including relevant places to modify when updating to net8.0+

AraHaan commented:
On the comment will you need to update those files in the PR as well (the ones mentioned on it).

BenVillalobos commented:
@AraHaan Yep!

AraHaan commented on Directory.Build.props:
```diff
@@ -38,7 +50,7 @@
         SYSLIB0011: Removing binary formatter will happen as part of a larger .NET-wide effort.
     -->
```
Add SYSLIB0037, SYSLIB0044, RS0016, and RS0017?

rainersigwald commented on src/StringTools/StringTools.csproj:
```diff
@@ -6,7 +6,7 @@
     <IsPackable>true</IsPackable>
     <GenerateReferenceAssemblySource>true</GenerateReferenceAssemblySource>
-    <LangVersion>8.0</LangVersion>
+    <LangVersion>11.0</LangVersion>
```
Should this be dropped or `latest`?

rainersigwald replied:
If it's just there temporarily to work around the compiler bug we can leave it, but I don't know of a great reason to explicitly specify an old langver . . .

BenVillalobos replied:
11 was the workaround because that's what part of the error message gave, but we can move this to latest and never worry about it again (🤞)

rainersigwald commented on src/Build.OM.UnitTests/ObjectModelRemoting/RemoteProjectsProviderMock/CollectionsHelpers.cs:
```diff
@@ -85,13 +85,13 @@ public static IList<MockProjectElementLinkRemoter> ExportCollection<T>(this Proj
 
 
-        public static IDictionary<key, T> ImportDictionary<key, T, RMock>(this ProjectCollectionLinker importer, IDictionary<key, RMock> source)
+        public static IDictionary<Key, T> ImportDictionary<Key, T, RMock>(this ProjectCollectionLinker importer, IDictionary<Key, RMock> source)
```
What motivated this change? Should it be `TKey` instead?

BenVillalobos replied:
The issue was "don't declare types like this as lowercase" so I just uppercased it. If TKey is what we should, thumbs up and I can make that change

rainersigwald replied:
Yeah, I agree with this guideline: https://stackoverflow.com/a/1098058

rainersigwald commented on src/StringTools/InternableString.Simple.cs:
```diff
@@ -51,7 +51,7 @@ namespace Microsoft.NET.StringTools
             private int _charIndex;
 
-            public Enumerator(ref InternableString spanBuilder)
+            public Enumerator(scoped ref InternableString spanBuilder)
```
Link to JaredPar's explanation on this since it's a workaround for a compiler bug.

lbussell commented:
Hi @BenVillalobos, what's the status of this? Can it be merged soon? Thanks!

rainersigwald commented:
This is not currently planned, because it breaks an SDK invariant (SDK tools should target LTS .NET). We have a meeting coming up to discuss a path forward with various stakeholders.

lbussell commented:
@rainersigwald Please consider adding the source-build team to the meeting if it's appropriate (myself, @crummel, @MichaelSimons). We were hoping to get this change in for RC2.

AraHaan commented:
> This is not currently planned, because it breaks an SDK invariant (SDK tools should target LTS .NET). We have a meeting coming up to discuss a path forward with various stakeholders.

I think it should be on for the .NET 7 SDK only, but yeah I would be ok with it targeting LTS for LTS sdk builds.

lbussell commented:
@BenVillalobos @rainersigwald @jaredpar, from the discussion we had in the meeting, is the plan now to merge this to flush out any potential issues going forwards?

rainersigwald commented:
Yes, this is now The Plan again. ETA next week, hopefully early.

rainersigwald commented on Directory.Build.props:
```diff
@@ -36,9 +48,11 @@
     -->
 
-    <NoWarn>$(NoWarn);NU1603;NU5105;1701;1702;SYSLIB0011</NoWarn>
+    <NoWarn>$(NoWarn);NU1603;NU5105;1701;1702;SYSLIB0011;SYSLIB0037;SYSLIB0044;RS0016;RS0017;</NoWarn>
```
Putting `RS0016` and `RS0017` on this list makes me _really_ nervous, since it means we'll miss breaking API changes. Maybe we should consider switching to the [package baseline validator](https://docs.microsoft.com/dotnet/fundamentals/package-validation/baseline-version-validator) instead?

rainersigwald reviewed (approved):
@BenVillalobos can you give this a once-over again? I made a couple of small changes:

* merged main
* updated to sdk rc1
* Undid the Roslyn-bug workaround since it's fixed in RC1
* made the style `TKey` change
* Dealt with a couple of new deprecation warnings

rainersigwald commented:
Internal official build for good measure https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6700270&view=results

BenVillalobos commented:
Potential workaround for our roslyn analyzer issue: https://github.com/dotnet/msbuild/issues/7903

rainersigwald commented:
@lbussell I suspect sourcebuild will also want us on Arcade 7 and I'm working on that migration now.
